libflux/flux-core/src/semantic/nodes.rs

```diff
@@ -867,6 +867,10 @@ impl FunctionExpr {
         }
         // And use it to infer the body.
         let (nenv, bcons) = self.body.infer(nenv, f)?;
+        // Update the argument variable with the types inferred from the body
+        for (k, t) in req.iter_mut() {
+            *t = nenv.lookup(k).unwrap().expr.clone();
+        }
```

Comment:
I think you need to do the same with the optional arguments as well as the required ones. I can think of an example where I think this would matter:
```js
apply = (a=(v) => v, b) => {
  f = a
  return f(v: b) + 1
}
```
In this case the type of parameter `a` would be `(v: X) => X` without considering uses in the body, but we need to further constrain the type of `a` to be `(v: int) => int`. Thoughts?